src/rgw/rgw_rest_s3.cc

```diff
@@ -6054,9 +6066,9 @@ int RGWS3Select::send_response_data(bufferlist &bl, off_t ofs, off_t len)
     m_previous_line = true;
   }
 
-  run_s3select((char *)query.c_str(), (char *)bl.c_str(),len,skip_first_line, m_previous_line /*skip last line*/, (m_processed_bytes >= s->obj_size));
+  status = run_s3select((char *)query.c_str(), (char *)bl.c_str(), len, skip_first_line, m_previous_line /*skip last line*/, (m_processed_bytes >= s->obj_size));
```

Comment:
yuval mentioned const_cast;  that said, could the called method not take a const query and perhaps bl arguments here?  second, could query be of type (const?) string& or (const?) string_view?  (clearly, this would imply shifting to c++/raii style coding in the new code being called.  So my attention was drawn to bl.c_str() in this call, and didn't at first notice it earlier in the enclosing method;  With respect to both, bl.c_str() is potentially suspicious, as if bl has multiple buffer segments (is "iovec" access pattern), then calling c_str() on it will copy all the segments into a new contiguous buffer.  if it's possible to avoid it, we save some thrashing

Reply:
in case bl contain multiple segments (thus , on call to c_str() , it copies ...) , this **must be avoided**.
the current technique of merging broken lines (on cross chunks cases) , should apply to bl segments.
**run_s3select** signature is not correct , the call stack (send_response --> breaking into lines --> breaking into tokens--> binding query variables to input --> execution on query variables) is about to change to avoid non-necessary copies and allocations/de-allocations.
the query and bl (input data) are not modified by this stack, it should be read-only.

this is will be fixed along with the modification to memory management (as mentioned) , the idea is to allocate once per query life cycle (instead of intensive new/delete per row/column).